File: domain/src/main/java/com/mrp/sml/domain/repository/DeviceConnectionRepository.java
package com.mrp.sml.domain.repository;

import com.mrp.sml.domain.model.DeviceInfo;
import java.util.List;

public interface DeviceConnectionRepository {
    ConnectionState getCurrentConnectionState();

    void observeConnectionState(ConnectionStateListener listener);
    void removeConnectionStateObserver(ConnectionStateListener listener);

    void observeDiscoveredDevices(DiscoveredDevicesListener listener);
    void removeDiscoveredDevicesObserver(DiscoveredDevicesListener listener);

    void discoverDevices();
    void connectToDevice(String deviceId);
    void disconnect();

    void performHandshake(String deviceAddress);
    void setDeviceInfo(DeviceInfo deviceInfo);
    DeviceInfo getConnectedDeviceInfo();

    interface ConnectionStateListener {
        void onConnectionStateChanged(ConnectionState state);
    }

    interface DiscoveredDevicesListener {
        void onDevicesUpdated(List<DiscoveredDevice> devices);
    }
}

File: domain/src/main/kotlin/com/mrp/sml/domain/repository/DeviceConnectionRepository.kt
package com.mrp.sml.domain.repository

import kotlinx.coroutines.flow.Flow

interface DeviceConnectionRepository {
    fun observeConnectionState(): Flow<ConnectionState>

    suspend fun discoverDevices()

    suspend fun connectToDevice(deviceId: String)

    suspend fun disconnect()
}

enum class ConnectionState {
    IDLE,
    DISCOVERING,
    CONNECTED,
    FAILED,
    DISCONNECTED,
}
